File: smart_trip_ai_mobile/lib/services/pedometer_engine.dart
import 'dart:math';

class PedometerResult {
  final int stepsCount;
  final double caloriesBurned;
  final bool stepDetected;

  PedometerResult({
    required this.stepsCount,
    required this.caloriesBurned,
    required this.stepDetected,
  });
}

class PedometerSummary {
  final int stepsCount;
  final double caloriesBurned;

  PedometerSummary({
    required this.stepsCount,
    required this.caloriesBurned,
  });
}

/// Real-time Physical Device Pedometer & Step Detection Engine.
/// Uses DeviceMotionEvent (Accelerometer peak detection algorithm) to count actual physical footsteps.
class PedometerEngine {
  int stepsCount = 0;
  double caloriesBurned = 0;
  int lastStepTimestamp = 0;
  double lastAccelMag = 9.81;
  final double stepThreshold = 11.2; // Vertical acceleration step impact threshold (m/s²)
  final int stepRefractoryMs = 350; // Minimum time between human footsteps (ms)

  void reset() {
    stepsCount = 0;
    caloriesBurned = 0;
    lastStepTimestamp = 0;
    lastAccelMag = 9.81;
  }

  /// Process 3-axis accelerometer vector from phone's hardware motion sensor
  PedometerResult processMotion(double accelX, double accelY, double accelZ) {
    final now = DateTime.now().millisecondsSinceEpoch;
    // Combined 3D acceleration magnitude (m/s²)
    final mag = sqrt(accelX * accelX + accelY * accelY + accelZ * accelZ);

    bool stepDetected = false;

    // Peak detection: Transition from low acceleration to impact peak above threshold
    if (mag > stepThreshold &&
        lastAccelMag <= stepThreshold &&
        (now - lastStepTimestamp) > stepRefractoryMs) {
      stepsCount += 1;
      // Approx 0.045 kcal per footstep
      caloriesBurned = double.parse((stepsCount * 0.045).toStringAsFixed(1));
      lastStepTimestamp = now;
      stepDetected = true;
    }

    lastAccelMag = mag;

    return PedometerResult(
      stepsCount: stepsCount,
      caloriesBurned: caloriesBurned,
      stepDetected: stepDetected,
    );
  }

  /// Fallback step estimation from real-time GPS distance when motion sensors are disabled
  PedometerSummary updateFromGpsDistance(double distanceKm) {
    final estimatedSteps = (distanceKm * 1310).round(); // ~1310 steps / km avg stride
    if (estimatedSteps > stepsCount) {
      stepsCount = estimatedSteps;
      caloriesBurned = double.parse((stepsCount * 0.045).toStringAsFixed(1));
    }
    return PedometerSummary(
      stepsCount: stepsCount,
      caloriesBurned: caloriesBurned,
    );
  }

  PedometerSummary getSummary() {
    return PedometerSummary(
      stepsCount: stepsCount,
      caloriesBurned: caloriesBurned,
    );
  }
}

final pedometerEngine = PedometerEngine();

File: src/services/pedometerEngine.ts
/**
 * Real-time Physical Device Pedometer & Step Detection Engine.
 * Uses DeviceMotionEvent (Accelerometer peak detection algorithm) to count actual physical footsteps.
 */

export class PedometerEngine {
  private stepsCount: number = 0;
  private caloriesBurned: number = 0;
  private lastStepTimestamp: number = 0;
  private lastAccelMag: number = 9.81;
  private stepThreshold: number = 11.2; // Vertical acceleration step impact threshold (m/s²)
  private stepRefractoryMs: number = 350; // Minimum time between human footsteps (ms)

  public reset(): void {
    this.stepsCount = 0;
    this.caloriesBurned = 0;
    this.lastStepTimestamp = 0;
    this.lastAccelMag = 9.81;
  }

  /**
   * Process 3-axis accelerometer vector from phone's hardware motion sensor
   */
  public processMotion(accelX: number, accelY: number, accelZ: number): {
    stepsCount: number;
    caloriesBurned: number;
    stepDetected: boolean;
  } {
    const now = Date.now();
    // Combined 3D acceleration magnitude (m/s²)
    const mag = Math.sqrt(accelX * accelX + accelY * accelY + accelZ * accelZ);

    let stepDetected = false;

    // Peak detection: Transition from low acceleration to impact peak above threshold
    if (
      mag > this.stepThreshold &&
      this.lastAccelMag <= this.stepThreshold &&
      now - this.lastStepTimestamp > this.stepRefractoryMs
    ) {
      this.stepsCount += 1;
      // Approx 0.045 kcal per footstep
      this.caloriesBurned = parseFloat((this.stepsCount * 0.045).toFixed(1));
      this.lastStepTimestamp = now;
      stepDetected = true;
    }

    this.lastAccelMag = mag;

    return {
      stepsCount: this.stepsCount,
      caloriesBurned: this.caloriesBurned,
      stepDetected,
    };
  }

  /**
   * Fallback step estimation from real-time GPS distance when motion sensors are disabled
   */
  public updateFromGpsDistance(distanceKm: number): {
    stepsCount: number;
    caloriesBurned: number;
  } {
    const estimatedSteps = Math.round(distanceKm * 1310); // ~1310 steps / km avg stride
    if (estimatedSteps > this.stepsCount) {
      this.stepsCount = estimatedSteps;
      this.caloriesBurned = parseFloat((this.stepsCount * 0.045).toFixed(1));
    }
    return {
      stepsCount: this.stepsCount,
      caloriesBurned: this.caloriesBurned,
    };
  }

  public getSummary(): { stepsCount: number; caloriesBurned: number } {
    return {
      stepsCount: this.stepsCount,
      caloriesBurned: this.caloriesBurned,
    };
  }
}

export const pedometerEngine = new PedometerEngine();
